src/context/auth.tsx

```diff
@@ -0,0 +1,142 @@
+import React, {
+  useReducer,
+  useEffect,
+  useContext,
+  PropsWithChildren,
+} from "react";
+import { checkAuth, signIn, signOut, signUp } from "../api";
+
+type IUser = {
+  id: string;
+  name: string;
+};
+
+interface IAuthState {
+  loading: boolean;
+  user: IUser | null;
+}
+
+interface IAuthAction {
+  type: "LOADING" | "AUTH" | "SIGN_OUT";
+  user?: IUser;
+}
+
+interface IAuthContext extends IAuthState {
+  signup: (username: string, password: string) => Promise<any>;
+  signin: (username: string, password: string) => Promise<any>;
+  signout: () => Promise<any>;
+}
+
+const reducer = (state: IAuthState, action: IAuthAction): IAuthState => {
+  if (action.type === "LOADING") {
+    return {
+      ...state,
+      loading: true,
+    };
+  }
+
+  if (action.type === "AUTH") {
+    return {
+      ...state,
+      user: action.user || null,
+      loading: false,
+    };
+  }
+
+  if (action.type === "SIGN_OUT") {
+    return {
+      ...state,
+      user: null,
+    };
+  }
+
+  return state;
+};
+
+const AuthContext = React.createContext<IAuthContext>({
+  user: null,
+  loading: true,
+  signup: (username: string, password: string) => Promise.resolve(undefined),
+  signin: (username: string, password: string) => Promise.resolve(undefined),
+  signout: () => Promise.resolve(undefined),
+});
+
+export const useAuth = () => {
+  const auth = useContext(AuthContext);
+
+  if (!auth) {
+    throw new Error("Must use within AuthProvider");
+  }
+
+  return auth;
+};
+
+const useIsAuthenticated = () => {
+  const { user } = useAuth();
+
+  return user !== null;
+};
+
+interface IAuthenticated {
+  fallback: JSX.Element;
+}
+
+export const Authenticated = (props: PropsWithChildren<IAuthenticated>) => {
+  const authenticated = useIsAuthenticated();
+
+  if (authenticated && React.isValidElement(props.children)) {
+    return props.children;
+  }
+
+  if (props.fallback) {
+    return props.fallback;
+  }
+
+  return null;
+};
+
+const AuthProvider = (props: React.PropsWithChildren<unknown>) => {
+  const [state, dispatch] = useReducer(reducer, { user: null, loading: true });
```

Comment:
I use a reducer here since, the `user` and `loading` state tend to change together. It makes sense to use grouped state when code starts looking like this.

```
setUser(user);
setLoading(fase);
```

Here I would use a reducer or `setState({ user, loading: false})`.